spoc/query_engine.py

```diff
@@ -172,83 +218,375 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
+        if self._function == AggregationFunction.COUNT:
+            aggregation_string = (
+                f"COUNT(*) as {self._value_column}_{self._function.name.lower()}"
+            )
+        elif self._function == AggregationFunction.AVG_WITH_EMPTY:
+            # For average, we need to sum up the values and divide by the number of regions
+            aggregation_string = f"SUM({self._value_column})::float/{input_schema.get_region_number()} as {self._value_column}_{self._function.name.lower()}"
+        else:
+            aggregation_string = f"{self._function.name}({self._value_column}) as {self._value_column}_{self._function.name.lower()}"
+        data_frame = (
+            data_frame.set_alias("data")
+            .aggregate(
+                ",".join(offset_columns + [aggregation_string]),
+            )
+            .order(",".join(offset_columns))
+        )
+        return data_frame
 
+    def _create_empty_dense_output(
+        self,
+        windowsize: int,
+        input_schema: GenomicDataSchema,
+        position_fields: Dict[int, List[str]],
+    ) -> duckdb.DuckDBPyRelation:
+        """Create dense value columns for all offsets."""
+        binsize: Optional[int] = input_schema.get_binsize()
+        if binsize is None:
+            raise ValueError("No binsize specified in data schema.")
+        int_binsize: int = binsize
+        # create combinations of offsets
+        offset_combinations = pd.DataFrame(
+            product(
+                np.arange(
+                    -(np.floor(windowsize / int_binsize) * int_binsize),
+                    (np.floor(windowsize / int_binsize) * int_binsize) + 1,
+                    int_binsize,
+                ),
+                repeat=len(position_fields.keys()),
+            ),
+            columns=[f"offset_{i}" for i in position_fields.keys()],
+        )
+        # fill value
+        if self._function in (
+            AggregationFunction.COUNT,
+            AggregationFunction.SUM,
+            AggregationFunction.AVG_WITH_EMPTY,
+        ):
+            offset_combinations["fill_value"] = 0
+        else:
+            offset_combinations["fill_value"] = np.nan
+        return duckdb.from_df(offset_combinations, connection=DUCKDB_CONNECTION)
 
-class Transformation:
-    """Transformation of contacts or pixels"""
+    def _fill_empty_output(
+        self,
+        data_frame: duckdb.DuckDBPyRelation,
+        empty_dense_output: duckdb.DuckDBPyRelation,
+        position_fields: Dict[int, List[str]],
+    ) -> duckdb.DuckDBPyRelation:
+        """Fill empty output with values from dense output."""
+        # get offset columns
+        offset_columns = [f"offset_{i}" for i in position_fields.keys()]
+        # construct join and coalesce output
+        data_frame = (
+            data_frame.set_alias("data")
+            .join(
+                empty_dense_output.set_alias("empty_dense_output"),
+                ",".join(offset_columns),
```

Comment:
If distances (offsets) from regions to fragments are in nts (not binned), how the join with binned empty offset df works?

Reply:
See answer to comment, but in short, no onctacts allowed for no